embedded-graphics/src/primitives/ellipse.rs

```diff
@@ -0,0 +1,512 @@
+        let inner_size = outer_size.saturating_sub(Size::new(2 * stroke_width, 2 * stroke_width));
+        let inner_size_sq = Size::new(inner_size.width.pow(2), inner_size.height.pow(2));
+        let iter = if !styled.style.is_transparent() {
+            Points::new(&styled.primitive)
+        } else {
+            Points::empty()
+        };
+
+        let threshold = compute_threshold(inner_size_sq, inner_size.width);
+
+        Self {
+            iter,
+            outer_color: styled.style.stroke_color,
+            inner_size_sq,
+            inner_color: styled.style.fill_color,
+            center: styled.primitive.center_2x(),
+            threshold,
+        }
+    }
+}
+
+impl<C> Iterator for StyledEllipseIterator<C>
+where
+    C: PixelColor,
+{
+    type Item = Pixel<C>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        for point in &mut self.iter {
+            let inside_border = is_point_inside_ellipse(
+                self.inner_size_sq,
+                point * 2 - self.center,
+                self.threshold,
+            );
+
+            let color = if inside_border {
+                self.inner_color
+            } else {
+                self.outer_color
+            };
+
+            if let Some(color) = color {
+                return Some(Pixel(point, color));
+            }
+        }
+
+        None
+    }
+}
+
+impl<'a, C: 'a> Drawable<C> for &Styled<Ellipse, PrimitiveStyle<C>>
+where
+    C: PixelColor,
+{
+    fn draw<D: DrawTarget<C>>(self, display: &mut D) -> Result<(), D::Error> {
+        display.draw_ellipse(self)
+    }
+}
+
+fn compute_threshold(size: Size, diameter: u32) -> u32 {
+    let Size {
+        width: a,
+        height: b,
+    } = size;
+
+    // Special case for circles, where width and height are equal
+    if a == b {
+        if diameter <= 4 {
+            diameter.pow(2) - diameter / 2
+        } else {
+            diameter.pow(2)
+        }
+    } else {
+        b * a
+    }
+}
+
+/// Uses the ellipse equation b^2 * x^2 + a^2 * y^2 - a^2 * b^2 to return a value signifying whether
+/// a given point lies inside (`true`) or outside (`false`) an ellipse centered around `(0, 0)` with
+/// width and height defined by the `size` parameter.
+fn is_point_inside_ellipse(size: Size, point: Point, threshold: u32) -> bool {
+    let Size {
+        width: a,
+        height: b,
+    } = size;
+
+    let Point { x, y } = point;
+
+    let x = x.pow(2) as u32;
+    let y = y.pow(2) as u32;
+
+    // Special case for circles, where width and height are equal
+    if a == b {
+        (x + y) < threshold
+    } else {
+        b * x + a * y < threshold
+    }
+}
+
+impl<'a, C> IntoIterator for &'a Styled<Ellipse, PrimitiveStyle<C>>
+where
+    C: PixelColor,
+{
+    type Item = Pixel<C>;
+    type IntoIter = StyledEllipseIterator<C>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        StyledEllipseIterator::new(self)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
If I haven't missed something, all tests seem to use `top_left = Point::zero()`. IMO at least one should use a different value for `top_left`.
This could maybe be combined with a test for `translate`, which isn't tested yet.

Reply:
I've added a combined non-zero origin and translate test.